_posts/2016-11-14-JS-Glossary.md

```diff
@@ -0,0 +1,554 @@
+
+### ES2016
+
+It took forever for [ES6](#es6) to come out, and every time they created / amended a specification there were multiple implementations of the specification available for transpiling via [babel](#babel). This I can imagine was frustrating for developers wanting to use new features, and specification authors trying to put out documentation for discussion as a work in progress. This happened a lot [with the Promises](#promises) API.
+
+To fix this they opted to discuss specification features on a year basis. So that specifications could be smaller and more focused, instead of major multi-year projects. Quite a SemVer jump from 6 to 2016.
+
+### Stages
+
+Turns out that didn't work out too well, so the terminology changed again. The change is mainly to set expectations between the Specification authors and developers transpiling those specifications into their apps.
+
+Now an ECMAScript language improvement specification moves through a series of stages, depending on their maturity. I believe starting at 4, and working down to 1. So a ECMAScript Stage 4 feature is going to be really new, if you're using it via a transpiler then you should expect a lot of potential API changes and code churn. The lower the number, the longer the spec has been discussed, and the more likely for the code you're transpiling to be the vanilla JavaScript code in time.
+
+
+### Modules / Imports
+
+A modules is the terminology for a group of JavaScript code. Terminology can get confusing, as the import structure for a library is very similar to importing a local file.
+
+You can import a module using syntax like `import { thin, other } from "thingy"`. Here's some examples [from our project][imports]:
+
+```js
+// Import modules
+import Relay from 'react-relay'
+import React from 'react'
+// Import two items from the react-native module 
+import { View, TouchableWithoutFeedback } from 'react-native'
+
+// Import the default class from a local file
+import ImageView from '../../opaque_image_view'
+import SwitchBoard from '../../../native_modules/switch_board'
+```
+
+An import can either have [a default export][default-export], or a set of exportable function/objects.
+
+You might see something like `const thing = require "thing"` around the internet, I think that's the older syntax from before the above syntax was decided upon.
+
+### Classes
+
+Modern JavaScript has classes introduced in [es6](#es6), this means that instead of writing something like:
+
+```js
+const danger = {
+  name: "Danger",
+  hello: function () {
+  console.log("Hi!")
+  }
+}
+
+danger.hello();
+```
+
+Instead you could write:
+
+```js
+class Person {
+  constructor(name) {
+    this.name = name
+  }
+}
+
+const danger = new Person("danger")
+```
+
+Classes provides the option of doing object-oriented programming, which is still a pretty solid way to write code. You can get useful errors 
```

Comment:
Might be worth mentioning Javascript's original OO paradigm called "prototypal inheritance". [This is a more opinionated article](https://medium.com/javascript-scene/the-two-pillars-of-javascript-ee6f3281e7f3#.tr1gwlz11), but is has an interesting section about prototypes vs. classes.